Module 5,Session 2  
Activation Functions and Forward Pass

In [1]:
import numpy as np

Exercise 1- Activation Functions

We are given the following inputs:
z = [-5, -1, 0, 1, 5]

Below are the outputs for different activation functions.

In [2]:
z = np.array([-5, -1, 0, 1, 5])

sigmoid = 1 / (1 + np.exp(-z))
tanh = np.tanh(z)
relu = np.maximum(0, z)
leaky_relu = np.where(z > 0, z, 0.1 * z)
swish = sigmoid * z

sigmoid, tanh, relu, leaky_relu, swish

(array([0.00669285, 0.26894142, 0.5       , 0.73105858, 0.99330715]),
 array([-0.9999092 , -0.76159416,  0.        ,  0.76159416,  0.9999092 ]),
 array([0, 0, 0, 1, 5]),
 array([-0.5, -0.1,  0. ,  1. ,  5. ]),
 array([-0.03346425, -0.26894142,  0.        ,  0.73105858,  4.96653575]))

Sigmoid
Outputs values between 0 and 1
Not zero-centered

Tanh
Outputs values between -1 and 1
Zero-centered

ReLU
Outputs 0 for negative values
Very computationally simple

Leaky ReLU
Like ReLU but allows small negative values

Swish
Smooth, modern activation
Used in advanced neural networks

Answers
Zero-centered output - Тanh
Computationally simplest - ReLU
Preferred for deep neural networks - ReLU / Leaky ReLU / Swish

Exercise 2 - Forward Pass

Input X = [0.5, 0.3]

Weights W1 =[[0.2, 0.4],[0.7, 0.9]]
Bias B1 = [0.1, 0.2]

Weights W2 =[[0.6],[0.8]]
Bias B2 = [0.3]

Step 1 - Hidden Layer Weighted Sum (Z1)

Z1 = X * W1 + B1

Z1₁ = (0.5 * 0.2) + (0.3 * 0.7) + 0.1  
Z1₁ = 0.1 + 0.21 + 0.1 = 0.41

Z1₂ = (0.5 * 0.4) + (0.3 * 0.9) + 0.2  
Z1₂ = 0.2 + 0.27 + 0.2 = 0.67

Z1 = [0.41, 0.67]

Step 2 - Apply ReLU

ReLU(x) = max(0, x)
A1 = [0.41, 0.67]

Step 3 - Output Layer Weighted Sum (Z2)

Z2 = A1 * W2 + B2

Z2 = (0.41 * 0.6) + (0.67 * 0.8) + 0.3  
Z2 = 0.246 + 0.536 + 0.3 = 1.082

Step 4 -  Sigmoid Activation
ŷ ≈ 0.747

Final Output - The network predicts approximately = 0.75

Exercise 3 - Architecture Design

1.Input Layer neurons  
32 × 32 = 1024 neurons  
Because each pixel is one input feature.

2.Hidden Layer activation
ReLU
It is computationally efficient and helps avoid vanishing gradients.

3.Output Layer neurons 
3 neurons
One for each class: Cat, Dog, Bird.

4.Output Layer activation
Softmax
Because this is a multi-class classification problem and Softmax outputs probabilities
that sum to 1.